pkg/provisioner/ironic/testserver/redfish.go

```diff
@@ -0,0 +1,703 @@
+package testserver
+
+import (
+	"net/http"
+	"testing"
+)
+
+// RedfishMock is a test server that implements Redfish Virtual Media semantics
+type RedfishMock struct {
+	*MockServer
+}
+
+// NewRedfish builds a redfish virtual media mock server
+func NewRedfish(t *testing.T) *RedfishMock {
+	return &RedfishMock{
+		MockServer: NewTLS(t, "redfish"),
+	}
+}
+
+// Endpoint returns the URL to the server
+func (m *RedfishMock) Endpoint() string {
+	return m.MockServer.RawEndpoint()
+}
+
+// RootService configures the server with a valid response for /redfish/v1/
+func (m *RedfishMock) rootService() *RedfishMock {
+	m.ResponseWithCode("/redfish/v1/", `
+	{
+		"@odata.type": "#ServiceRoot.v1_9_0.ServiceRoot",
+		"Id": "RootService",
+		"Name": "Root Service",
+		"RedfishVersion": "1.6.0",
+		"UUID": "92384634-2938-2342-8820-489239905423",
+		"Systems": {
+			"@odata.id": "/redfish/v1/Systems"
+		},
+		"Chassis": {
+			"@odata.id": "/redfish/v1/Chassis"
+		},
+		"Managers": {
+			"@odata.id": "/redfish/v1/Managers"
+		},
+		"Tasks": {
+			"@odata.id": "/redfish/v1/TaskService"
+		},
+		"SessionService": {
+			"@odata.id": "/redfish/v1/SessionService"
+		},
+		"AccountService": {
+			"@odata.id": "/redfish/v1/AccountService"
+		},
+		"EventService": {
+			"@odata.id": "/redfish/v1/EventService"
+		},
+		"Registries": {
+			"@odata.id": "/redfish/v1/Registries"
+		},
+		"UpdateService": {
+			"@odata.id": "/redfish/v1/UpdateService"
+		},
+		"CertificateService": {
+			"@odata.id": "/redfish/v1/CertificateService"
+		},
+		"Links": {
+			"Sessions": {
+				"@odata.id": "/redfish/v1/SessionService/Sessions"
+			}
+		},
+		"Oem": { },
+		"@odata.id": "redfish/v1/"
+	}`, http.StatusOK)
+	return m
+}
+
+// Unauthorized configures the server with a http 401 Unauthorized response for /redfish/v1/
+func (m *RedfishMock) Unauthorized() *RedfishMock {
+	m.ErrorResponse("/redfish/v1/", http.StatusUnauthorized)
+	return m
+}
+
+// NoSystem configures the server with a valid response for /redfish/v1
+func (m *RedfishMock) NoSystem() *RedfishMock {
+	m.rootService()
```

Comment:
That's not a recommended practice. The setup of the mock must be defined in the setup code portion of the test, using the fluent builder pattern. In general, every method of the mock interface must focus on mocking a single API call (there could be exceptions for that, but only for very generic behaviors).

Reply:
Sure, thanks!

Reply:
Basically gofish doesn't like much the idea of marshal/unmarshal from the struct I took a different approach to return the API calls in the tests.